### ISSUE TITLE
KafkaConsumer intermittent hanging

### ISSUE THREAD

**Issue by kenjones-cisco** [dpkp/kafka-python#1520]
#### Summary
Using a similar pattern as shown in the [example](https://github.com/dpkp/kafka-python/blob/1.4.3/example.py) for the consumer implementation, I notice that after a period of time there are intermittent hanging of the process.

Snippet from logs
```
2018-06-14 13:58:49,029.29.598712921142578:kafka.protocol.parser:140470596655848:DEBUG:14:Received correlation id: 2148
2018-06-14 13:58:49,029.29.874801635742188:kafka.protocol.parser:140470596655848:DEBUG:14:Processing response FetchResponse_v4
2018-06-14 13:58:49,030.30.13777732849121:kafka.conn:140470596655848:DEBUG:14:<BrokerConnection node_id=1 host=msg-brk-dc3-001.cisco.com:9093 <connected> [IPv4 ('64.101.96.18', 9093)]> Response 2148 (502.3922920227051 ms): FetchResponse_v4(throttle_time_ms=0, topics=[(topics='mcmp.openstackrequest.dev', partitions=[(partition=0, error_code=0, highwater_offset=1300, last_stable_offset=-1, aborted_transactions=NULL, message_set=b'')])])
2018-06-14 13:59:56,680.680.2170276641846:kafka.client:140470597716712:DEBUG:14:Sending metadata request MetadataRequest_v1(topics=NULL) to node 1
2018-06-14 13:59:56,680.680.4563999176025:kafka.protocol.parser:140470597716712:DEBUG:14:Sending request MetadataRequest_v1(topics=NULL)
``` 
More logs can be seen here: https://gist.github.com/kenjones-cisco/840335443287f13a88c6ae5a02d504f7

Looking at the timestamps, there is a just over 7 seconds where it hangs. During this period of several minutes there was no messages received or processed. As such the consumer was idle and just polling when this happens.

#### Version

	kafka-python       1.4.3

#### Configurations

	client_id == container `HOSTNAME`
	group_id set
	bootstrap_servers set
	auto_offset_reset = 'earliest'
	consumer_timeout_ms = 240000
	// all others are using the defaults

Looking to verify if this is a bug or possibly a configuration error.

**Comment by kenjones-cisco** [comment]
I updated the gist with another file that shows another hang that is similar but at a different point, that lasts much longer.

Snippet from logs (other-hang-logs.txt)
```
2018-06-14 15:52:23,248.248.65150451660156:kafka.protocol.parser:140470596655848:DEBUG:14:Received correlation id: 14321
2018-06-14 15:52:23,248.248.89731407165527:kafka.protocol.parser:140470596655848:DEBUG:14:Processing response FetchResponse_v4
2018-06-14 15:52:23,249.249.17316436767578:kafka.conn:140470596655848:DEBUG:14:<BrokerConnection node_id=1 host=msg-brk-dc3-001.cisco.com:9093 <connected> [IPv4 ('64.101.96.18', 9093)]> Response 14321 (501.8439292907715 ms): FetchResponse_v4(throttle_time_ms=0, topics=[(topics='mcmp.openstackrequest.dev', partitions=[(partition=0, error_code=0, highwater_offset=1306, last_stable_offset=-1, aborted_transactions=NULL, message_set=b'')])])
2018-06-14 15:52:34,960.960.2408409118652:kafka.client:140470597716712:DEBUG:14:Sending metadata request MetadataRequest_v1(topics=['mcmp.openstackresponse.dev']) to node 1
2018-06-14 15:52:34,960.960.486650466919:kafka.protocol.parser:140470597716712:DEBUG:14:Sending request MetadataRequest_v1(topics=['mcmp.openstackresponse.dev'])
2018-06-14 15:52:34,960.960.878849029541:kafka.conn:140470597716712:DEBUG:14:<BrokerConnection node_id=1 host=msg-brk-dc3-001.cisco.com:9093 <connected> [IPv4 ('64.101.96.18', 9093)]> Request 35: MetadataRequest_v1(topics=['mcmp.openstackresponse.dev'])
2018-06-14 15:52:34,962.962.4395370483398:kafka.protocol.parser:140470597716712:DEBUG:14:Received correlation id: 35
2018-06-14 15:52:34,962.962.6002311706543:kafka.protocol.parser:140470597716712:DEBUG:14:Processing response MetadataResponse_v1
2018-06-14 15:52:34,962.962.9201889038086:kafka.conn:140470597716712:DEBUG:14:<BrokerConnection node_id=1 host=msg-brk-dc3-001.cisco.com:9093 <connected> [IPv4 ('64.101.96.18', 9093)]> Response 35 (2.187490463256836 ms): MetadataResponse_v1(brokers=[(node_id=2, host='msg-brk-dc3-002.cisco.com', port=9093, rack=None), (node_id=5, host='msg-brk-dc3-005.cisco.com', port=9093, rack=None), (node_id=4, host='msg-brk-dc3-004.cisco.com', port=9093, rack=None), (node_id=1, host='msg-brk-dc3-001.cisco.com', port=9093, rack=None), (node_id=3, host='msg-brk-dc3-003.cisco.com', port=9093, rack=None)], controller_id=4, topics=[(error_code=0, topic='mcmp.openstackresponse.dev', is_internal=False, partitions=[(error_code=0, partition=0, leader=1, replicas=[1, 4], isr=[4, 1])])])
2018-06-14 15:52:34,963.963.2868766784668:kafka.cluster:140470597716712:DEBUG:14:Updated cluster metadata to ClusterMetadata(brokers: 5, topics: 1, groups: 0)
2018-06-14 15:54:59,959.959.3510627746582:kafka.coordinator.consumer:140470716779336:DEBUG:14:Sending offset-commit request with {TopicPartition(topic='mcmp.openstackrequest.dev', partition=0): OffsetAndMetadata(offset=1306, metadata='')} for group mcmp to 5
2018-06-14 15:54:59,959.959.5980644226074:kafka.protocol.parser:140470716779336:DEBUG:14:Sending request OffsetCommitRequest_v2(consumer_group='mcmp', consumer_group_generation_id=1039, consumer_id='ansible-driver-34-8nnkl-59d028f2-f5bd-4d1d-9a67-9a6afae6dc6e', retention_time=-1, topics=[(topic='mcmp.openstackrequest.dev', partitions=[(partition=0, offset=1306, metadata='')])])
2018-06-14 15:54:59,960.960.1263999938965:kafka.conn:140470716779336:DEBUG:14:<BrokerConnection node_id=5 host=msg-brk-dc3-005.cisco.com:9093 <connected> [IPv4 ('64.101.96.22', 9093)]> Request 3940: OffsetCommitRequest_v2(consumer_group='mcmp', consumer_group_generation_id=1039, consumer_id='ansible-driver-34-8nnkl-59d028f2-f5bd-4d1d-9a67-9a6afae6dc6e', retention_time=-1, topics=[(topic='mcmp.openstackrequest.dev', partitions=[(partition=0, offset=1306, metadata='')])])
```

There are back to back hangs in this one. First like the first one I reported that happens from `2018-06-14 15:52:23,249.249` to `2018-06-14 15:52:34,960.960` (~11s), but then another one happens almost immediately from `2018-06-14 15:52:34,963` to `2018-06-14 15:54:59,959`(~2m25s).

**Comment by kngenie** [comment]
I've been fighting with frequent (like once a day) problem of `KafkaConsumer` hanging indefinitely. I think this issue is the closest, so I hang up my findings here.

I posted a gist with stack trace of two threads in deadlock: https://gist.github.com/kngenie/95f8f1a2e86a0ece587530d3b7ef4779 

In essence, main thread (app thread reading messages) holds `ConsumerCoordinator._lock`, Heatbeat thread holds `ConsumerCoordinator._client._lock`. Each is trying to acquire the other lock held by the other.

Root cause appears to me, that `ensure_active_group` is holing `self._lock` for possibly long period of time, even when it calls `self.ensure_coordinator_ready()`, `self._client.poll()` etc. that tries to acquire locks with `with self._client.lock, self._lock:`.

**Comment by jbf-zuora** [comment]
I'm pretty sure I have the same issue that @kngenie is seeing.

**Comment by bain2236** [comment]
I also have this issue. 
```python
consumer =KafkaConsumer(topic, group_id=group, bootstrap_servers=self.kafka_list, \
auto_off_reset='earliest', enable_auto_commit=True, auto_commit_interval_ms = 1000, \
consumer_timeout_ms=timeout, max_poll_interval_ms = 10000
)
```
My consumer is hanging inside the for message in consumer loop for periods longer than the max_poll which I've recently added to try and rectify the issue.
@kngenie Have you found a fix for this?

**Comment by kngenie** [comment]
@bain2236 I tried patching 1.4.3 to make `ConsumerCoordinator` release `self._lock` before everywhere it tries to acquire `_client._lock`, and it became far more stable. But code now looks so messy that I'm feeling bad with submitting a PR.  I hope author can find a cleaner way.
Also I'm still seeing a problem with heatbeat session that I suspect might be lock related as well.

**Comment by dpkp** [comment]
Have you tried 1.4.4? We made some further fixes to lock ordering.

**Comment by jeffwidman** [comment]
There were even further fixes on `master` that will be released as part of `1.4.5`.

So please try that (until a release happens, you can install latest `master` directly from github), and if you see hanging please file a new issue.
